Zib-AnatomicalLocation

ArdonToonstra opened the pull request:
Initial commit of new profiling proposal based on recent discussions (https://nictiz.atlassian.net/browse/ZIBFHIR2024-173).

pieter-edelman-nictiz commented on zib/StructureDefinitions/zib-AnatomicalLocation.xml:
```diff
@@ -0,0 +1,75 @@
+      <path value="CodeableConcept.coding" />
+      <binding>
+        <strength value="required" />
+        <valueSet value="http://decor.nictiz.nl/fhir/ValueSet/2.16.840.1.113883.2.4.3.11.60.121.11.11--20240415000000" />
```
I have some second thoughts on binding this ValueSet to `CodeableConcept.coding`. It precludes the use of additional codings, e.g. the following would be impossible.

```xml
<Condition xmlns="http://hl7.org/fhir">
    <meta>
        <profile value="https://example.org/fhir/StructureDefinition/Example-zib-Problem"/>
    </meta>
    <bodySite>
        <coding>
            <system value="http://snomed.info/sct"/>
            <code value="181923006"/>
        </coding>
        <coding>
            <system value="http://loinc.org"/>
            <code value="36023-0"/>
        </coding>
    </bodySite>   
    <subject>
        <reference value="Patient/1234"/>
        <type value="Patient"/>
    </subject>
</Condition>
```

In addition, it would result in a different way to define apply bindings than in the rest of the profiles.

So this makes me wonder: is the problem we're trying to solve really this big that we need to resort to a hackish solution? It's a problem for a select group of people that creates profiles. Couldn't we solve this in another way (perhaps a QA step?)

ArdonToonstra replied:
Indeed... this is not desired. We could slice `.coding`... make that slice 1..1 to account for the required binding and the option to add other codes. I think I forgot the slicing here.

I have also asked the question if/why we can't bind anything on the CodeableConcept level: https://chat.fhir.org/#narrow/stream/179166-implementers/topic/ValueSet.20binding.20on.20CodeableConcept.20in.20datatype.20profile 

If this turns out useless, we will add the ValueSet from the host profiles and provide documentation to do so in the profile like we did in 2020. Because I think you are right. It affects a small group of people that have to deal with this.

ArdonToonstra replied:
Moved in the direction of setting the binding on the CodeableConcept root.

pieter-edelman-nictiz commented on zib/StructureDefinitions/zib-AnatomicalLocation.xml:
```diff
@@ -0,0 +1,75 @@
+      <path value="CodeableConcept" />
+      <short value="Location" />
+      <definition value="Localisation on/in the body." />
+      <comment value="The zib AnatomicalLocation is mapped to both a profile on BodyStructure (&lt;http://nictiz.nl/fhir/StructureDefinition/zib-AnatomicalLocation-BodyStructure&gt;) and this CodeableConcept profile. The CodeableConcept handles the representation of the location, while the BodyStructure profile covers both location and laterality.&#xD;&#xA;&#xD;&#xA;In cases where only the locati ...
```
This comment should be placed in `StructureDefinition.purpose`, but terser comment should be placed here about when to use the one and when to use the other.

ArdonToonstra replied:
Moved some of this to the SD purpose and made this comment more concise.

ArdonToonstra commented:
Reworked the profiles by moving the location binding from `CodeableConcept.coding` to `CodeableConcept`. Also added all QA requirements, including release-notes based on a newly created public jira ticket (https://nictiz.atlassian.net/browse/ZIBFHIR-113)

 Tested this approach using a dummy profile and multiple examples:
- Forge does not remove the binding when opened in host profile (it does when you open and resave the AnatomicalLocation profile on CodeableConcept)
- No validator errors. 
- However, I was not able to trigger a validation error with incorrect values in in the host profile while having a Required binding in the datatype profile. This was both the case for the Java and .NET validator. This might require further investigation.

@pieter-edelman-nictiz  would you be willing to have another look?

pieter-edelman-nictiz commented:
The QA error is quite trivial; in the nl-core profile, you should override the mention of “zib-” profile to the “nl-core-” profile.

ArdonToonstra commented:
> When generating a snapshot in Forge or using the HL7 Validator, the binding from the dataset profile is ignored and reverts to the default from the base resource.

@pieter-edelman-nictiz , thanks for checking. I will investigate further... I posed this point on zulip. Might want to include this then to our list of things to check. If it doesn't work out in before publication time, we can revert back to the 'old' way and include the binding in the host profiles. Agree?

ArdonToonstra commented:
@pieter-edelman-nictiz / @LuudSlagter - could one of you review this PR?
Because of reasons mentioned in https://nictiz.atlassian.net/browse/ZIBFHIR2024-173 we revert back to using the ValueSet binding in the host profile. This has been processed in the profiles.

Moreover, I believe with the current profiling it is possible to only provide a laterality, if necessary. 

Lastly, I don't understand why the QA tooling is not triggered. Locally, the QA tool gives green.

pieter-edelman-nictiz commented on zib/StructureDefinitions/zib-AnatomicalLocation.xml:
```diff
@@ -0,0 +1,92 @@
+      <path value="CodeableConcept" />
+      <short value="Location" />
+      <definition value="Localisation on/in the body." />
+      <comment value="In cases where only the location is required, or where laterality is embedded within the location code, this CodeableConcept can be used within the resource. However, when laterality needs to be explicitly recorded, the BodyStructure resource should be used, referenced through the BodyStructure Reference within this CodeableConcept. If the BodyStructure reference is used, th ...
```
Please keep in mind dat this is the comment that ends up the profile when using this data type profile. It should be targeted at users of the profile.

ArdonToonstra replied:
This is text is aimed towards implementers of this profile. Not sure what of it is not. Could you highlight what you think is wrong?

ArdonToonstra replied:
maybe the 'required' part can be adjusted

ArdonToonstra replied:
Slightly adjusted.

pieter-edelman-nictiz replied:
So, I'm trying to imagine reading this text on e.g. `Observation.bodySite`.

> A deviation from our general profiling has been implemented in this profile.

From an implementer perspective, it might seem strange to see the words "our" and "profile" here. I think it would be more logical to have something like:

> In this case, slicing is not according to the Nictiz profiling guidelines.

Some other suggestions:

> Slicing has been applied to `CodeableConcept.coding` to enable the use of the `BodyStructure` reference extension without requiring a populated coded value. This approach facilitates scenarios where only the Laterality element needs to be conveyed.

Could be adjusted to:

> Slicing has been applied to `CodeableConcept.coding` instead of `CodeableConcept` directly to enable the use of the `BodyStructure` reference extension without requiring a coded value. This approach facilitates scenarios where only the zib Laterality concept needs to be conveyed.

And I think the following is not true:

> The decision was made due to the Location concept’s required binding strength. If the required binding were applied on the CodeableConcept, it would mandate a coded value. This slicing approach ensures flexibility while maintaining compliance with the required binding constraints for Location.

This way of slicing doesn't comply with the required binding constraint, as it makes the use codes from this value set completely optional.

ArdonToonstra replied:
Thanks those are indeed good textual improvements.

Regarding "This slicing approach ensures flexibility while maintaining compliance with the required binding constraints for Location." I was thinking of the level of where the mapping is placed, namely a .coding element with 0..1 with a required binding indicating that a code needs to be chosen from those valuesets. Together with the invariant it gets very close to a similar effect as a required binding strenght.  

But reading it again, I also agree that the text needs updating.

ArdonToonstra replied:
Outdated and removed with latest commit.

pieter-edelman-nictiz commented on zib/StructureDefinitions/zib-AnatomicalLocation.xml:
```diff
@@ -0,0 +1,92 @@
+        </discriminator>
+        <rules value="open" />
+      </slicing>
+      <comment value="A deviation from our general profiling has been implemented in this profile. Slicing has been applied to `CodeableConcept.coding` to enable the use of the `BodyStructure` reference extension without requiring a populated coded value. This approach facilitates scenarios where only the Laterality element needs to be conveyed.&#xA;&#xA;The decision was made due to the Location  ...
```
Unless I'm missing something, I don't think this slicing trick is going to help. From the zib perspective, the desired solution is to
- either use (at least) a code from the zib value set
- or use a reference to BodyStructure

Since the slice is optional, the use of a code from this value set is also optional -- any other code would be valid. Effectively, this has the same effect as an _extensible_ binding on the data type itself.

I guess an invariant could be used to satisfy the requirement, but that might be a bit overkill.

ArdonToonstra replied:
From the zib perspective the desired solution for the 'Location' concept is indeed to either have a code from the zib value sets here or in the BodyStructure resource via the reference. 

Normally in our FHIR profiling, if we bind a required value set, there is still the option to add additional codes (because Coding is 0..*).  This is still honored with this slicing trick.

The current slicing has indeed _almost_ the same effect as an extensible binding on the CodeableConcept. It saves adding the binding in the host profiles. ;) 

We could make an invariant that checks that if Codings are uses, alt least one Coding needs to have one of the systems from the ValueSet (http://snomed.info/sct and urn:oid:2.16.840.1.113883.6.43.1) which will help make sure it falls in the 0..1 slice for location? Not 100% watertight because a descendent-of is used for snomed...)

ArdonToonstra replied:
@pieter-edelman-nictiz  added an invariant - tested it and it seems to work with .NET and Java QA stuff. What do you think?

pieter-edelman-nictiz replied:
Nice work with the constraint!

But I'm not in favor of this slicing construct. It might make it easier for us but, but I think it makes it harder for profile consumers.

ArdonToonstra replied:
@pieter-edelman-nictiz  I don't see too much of a problem with it for profile consumers.

The only other option we have is to add the ValueSet to the CodeableConcept in the host profile with an Extensible binding instead of Required? Or am I missing something here?

ArdonToonstra replied:
This will result in a lot of entries in the known-issues.yml, to silence the zib-compliance checks. Correct?

ArdonToonstra replied:
Ok! Next proposal is with the extensible binding. I made the changes. 
The known-issue.yml will need to be enriched in the host profiles that use this I think. We will see.

pieter-edelman-nictiz commented on zib/StructureDefinitions/zib-AnatomicalLocation-BodyStructure.xml:
```diff
@@ -0,0 +1,77 @@
+    <element id="BodyStructure">
+      <path value="BodyStructure" />
+      <short value="AnatomicalLocation" />
+      <definition value="The zib AnatomicalLocation is mapped to both a profile on CodeableConcept (&lt;http://nictiz.nl/fhir/StructureDefinition/zib-AnatomicalLocation&gt;) and this BodyStructure profile. While the CodeableConcept profile covers anatomical location, the BodyStructure resource is used to capture both the location and its laterality.&#xA;&#xA;When laterality is required, this Body ...
```
I would remove the sentence "While the CodeableConcept profile covers anatomical location, the BodyStructure resource is used to capture both the location and its laterality". And then maybe adjust the next sentence a bit: "When it is required to convey a discrete Laterality (NL-CM:20.7.3) next to (or in absence of) the Location (NL-CM:20.7.4), this BodyStructure profile should be used and referenced from the relevant resource. In cases where Laterality is either unnecessary or embedded within the location code itself, it can be conveyed directly in the relevant resource."

ArdonToonstra replied:
Done! Thanks.

ArdonToonstra commented:
@pieter-edelman-nictiz good to go?

ArdonToonstra commented:
This profile is tested and reviewed in context of Zib Problem: https://github.com/Nictiz/Nictiz-R4-zib2024/pull/49

LuudSlagter commented:
This PR can only be merged after _OTH_ is removed from LateraliteitCodelijst.
